### PR TITLE
Use precomputed hash in probe side of HashSemiJoinOperator when available

### LINKED ISSUE #766: Use precomputed hash in HashSemiJoinOperator for probe
SemiJoin does not make use of precomputed hash on probe side when it is available. By default, `optimizer.optimize-hash-generation` is set to true so all SOURCE stages will output precomputed hash too. Ignoring it causes HashSemiJoinOperator to do duplicate computations to calculate hash for probe side and use it for joining.

Using the precomputed hash resulted in ~32% improvement in performance for below sample query on tpcds scale 1000:
`select count(*) from store_sales where store_sales.ss_customer_sk not in (select c_customer_sk from customer)`

I will open a pull request for these changes.

### PR DESCRIPTION
**Pull request by stagraqubole**
fixes #766

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/operator/HashSemiJoinOperator.java [hunk 1]
```diff
@@ -71,18 +75,24 @@ public void noMoreOperators()
...
 
     private final int probeJoinChannel;
     private final ListenableFuture<ChannelSet> channelSetFuture;
+    private final Optional<Integer> probeHashChannel;
 
     private ChannelSet channelSet;
     private Page outputPage;
     private boolean finishing;
 
     public HashSemiJoinOperator(OperatorContext operatorContext, SetSupplier channelSetFuture, int probeJoinChannel)
```
Comment (dain) [inline presto-main/src/main/java/io/prestosql/operator/HashSemiJoinOperator.java]: `HashSemiJoinOperator` is an internal detail of Presto, so we don't need to maintain API backwards compatibility.  This constructor is no longer used, so let's just remove it.

File: presto-main/src/main/java/io/prestosql/operator/HashSemiJoinOperator.java [hunk 2]
```diff
@@ -156,7 +167,14 @@ public void addInput(Page page)
                 }
             }
             else {
-                boolean contains = channelSet.contains(position, probeJoinPage);
+                boolean contains;
+                if (probeHashChannel.isPresent()) {
+                    long rawHash = BIGINT.getLong(page.getBlock(probeHashChannel.get()), position);
```
Comment (dain) [inline presto-main/src/main/java/io/prestosql/operator/HashSemiJoinOperator.java]: Maybe hoist `page.getBlock(probeHashChannel.get()` outside of the `for` loop:
```
Optional<Block> hashBlock = probeHashChannel.map(page::getBlock);
for (...) {
...
    if (hashBlock.isPresent()) {
        long rawHash = BIGINT.getLong(hashBlock.get(), position);
        contains = channelSet.contains(position, probeJoinPage, rawHash);
    }
    else {
...
````

**Comment by stagraqubole** [comment]
> Two minor comments, but otherwise looks good.
> 
> One last thing, can you change the commit message to just `Use precomputed hash in probe side of HashSemiJoinOperator`, so it doesn't wrap?

Done

**Comment by dain** [comment]
Merged. Thanks for the contribution!

**Pull request merged**